### PR TITLE
Implemented the ability to Stop Test Run if requested

### PR DESCRIPTION
**Pull request by gortegaMS**
# Type of PR

- [x] Documentation changes
- [x] Code changes
- [x] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## PR Checklist

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My code follows the code style of this project.
- [x] I ran lint checks locally prior to submission.
- [x] Have you checked to ensure there aren't other open Pull Requests for the same update/change?

## Does this introduce a breaking change

- [ ] YES
- [x] NO

## Validation

- [x] Unit tests updated and ran successfully
- [x] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes https://github.com/retaildevcrews/wcnp/issues/947
- Closes https://github.com/retaildevcrews/wcnp/issues/948

### CODE REVIEW

**Review by TravisFallsMSFT — APPROVED** [review]

File: docs/TestRunCancellation.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+- Starts Interval Checker to watch for an incoming `Cancellation Request`
+  - Creates LoadRunner command mode instance
+    - Upon completion, an execution exception or cancellation request will trigger `TestRun Complete` event to Post the TestRun current state back into CosmosDB.
+  - Ends LodeRunner command mode instance
+
+- Runs Retry HardStopTime checker to validate and logging `Cancellation Request completed` if was requested.
+
+### Logging
+
+- If a `Cancellation Request` is received during TestRun execution, a `TestRun Cancellation request received` message will be logged.
```
[the remark this answers is not in the document]
Comment (gortegaMS) [reply docs/TestRunCancellation.md:19]: This is a good point I added a link to it.

File: src/LodeRunner.Data/Services/TestRunService.cs [hunk 2]
```diff
@@ -72,8 +72,12 @@ public async Task<IEnumerable<TestRun>> GetNewTestRunsByLoadClientId(string load
             string sql = $"SELECT * FROM e WHERE e.entityType='TestRun' ";
             sql += $"and array_contains(e.loadClients, {{ \"id\": \"{loadClientId}\"}}, true) ";
```
[the remark this answers is not in the document]
Comment (gortegaMS) [reply src/LodeRunner.Data/Services/TestRunService.cs:73]: Created a triage item for this comment.

**Review by PurpleBriar — CHANGES_REQUESTED** [review]
Review failing github action.

File: src/LodeRunner.Data/CosmosDBRepository.cs [hunk 3]
```diff
@@ -202,10 +202,17 @@ public async Task<TEntity> CreateDocumentAsync<TEntity>(TEntity newDocument)
         /// <typeparam name="TEntity">The type of the entity.</typeparam>
         /// <param name="newDocument">The new document.</param>
         /// <param name="cancellationToken">The Cancellation Token.</param>
+        /// <param name="eTag">The entity tag associated with the resource.</param>
         /// <returns>An instance of the document.</returns>
-        public async Task<TEntity> UpsertDocumentAsync<TEntity>(TEntity newDocument, CancellationToken cancellationToken = default)
+        public async Task<TEntity> UpsertDocumentAsync<TEntity>(TEntity newDocument, CancellationToken cancellationToken = default, string eTag = "")
```
Comment (alfredoihernandez) [inline src/LodeRunner.Data/CosmosDBRepository.cs:207]: might be better to switch etag here for ItemRequestOptions

File: src/LodeRunner/src/Services/LodeRunnerService.cs [hunk 4]
```diff
@@ -230,44 +230,69 @@ public void LogStatusChange(object sender, ClientStatusEventArgs args)
...
-                    LoadClient = this.loadClient,
-                    StartTime = args.StartTime,
-                    ErrorMessage = args.ErrorMessage,
-                };
-
-                testRun.ClientResults.Add(loadResult);
-
-                // update TestRun CompletedTime if last client to report results
-                if (testRun.ClientResults.Count == testRun.LoadClients.Count)
+            DateTime dt = DateTime.Now;
```
Comment (alfredoihernandez) [inline src/LodeRunner/src/Services/LodeRunnerService.cs:245]: switch to utcnow

File: src/LodeRunner/src/ValidationTest/Main.cs [hunk 5]
```diff
@@ -258,8 +258,16 @@ public async Task<int> RunOnce(Config config, CancellationToken token)
             }
             finally
             {
-                // fire event
-                TestRunComplete(null, new LoadResultEventArgs(startTime, DateTime.UtcNow, config.TestRunId, requestCount, validationFailureCount + errorCount));
+                if (token.IsCancellationRequested)
+                {
+                    // fire event
+                    TestRunComplete(null, new LoadResultEventArgs(startTime, DateTime.UtcNow, config.TestRunId, 0, 0, SystemConstants.TestRunExecutionStoppedMessage));
```
Comment (alfredoihernandez) [inline src/LodeRunner/src/ValidationTest/Main.cs:264]: might be worth capturing requestcount and failure count inside cancellation event and leave it up to user/ui to determine how to use that data

File: src/LodeRunner/src/ValidationTest/Main.cs [hunk 6]
```diff
@@ -390,8 +398,15 @@ public int RunLoop(Config config, CancellationToken token)
                 totalRequests += state.Count;
             }
 
-            // fire event
-            TestRunComplete(null, new LoadResultEventArgs(startTime, completedTime, config.TestRunId, (int)totalRequests, totalFailures));
+            if (token.IsCancellationRequested)
+            {
+                TestRunComplete(null, new LoadResultEventArgs(startTime, DateTime.UtcNow, config.TestRunId, 0, 0, SystemConstants.TestRunExecutionStoppedMessage));
```
Comment (alfredoihernandez) [inline src/LodeRunner/src/ValidationTest/Main.cs:403]: might be worth capturing requestcount and failure count inside cancellation event and leave it up to user/ui to determine how to use that data

File: src/LodeRunner/src/Services/LodeRunnerService.cs [hunk 7]
```diff
@@ -230,44 +230,69 @@ public void LogStatusChange(object sender, ClientStatusEventArgs args)
...
+                    // Only set HardStopTime if all loadClients completed
+                    if (testRunResponse.Resource.HardStop && testRunResponse.Resource.HardStopTime == null)
+                    {
+                        testRunResponse.Resource.HardStopTime = DateTime.UtcNow;
+                    }
                 }
 
-                // post updates
-                _ = await GetTestRunService().Post(testRun, this.cancellationTokenSource.Token);
+                bool preconditionFailedExceptionThown = await TestRunExecutionHelper.TryCatchPreconditionFailedException(logger, nameof(UpdateTestRun), async () =>
```
Comment (alfredoihernandez) [inline src/LodeRunner/src/Services/LodeRunnerService.cs:271]: typo
```suggestion
                bool preconditionFailedExceptionThrown = await TestRunExecutionHelper.TryCatchPreconditionFailedException(logger, nameof(UpdateTestRun), async () =>
```

File: src/LodeRunner/src/Services/LodeRunnerService.cs (hunk 4 above)
Comment (gortegaMS) [reply src/LodeRunner/src/Services/LodeRunnerService.cs]: removed, not needed

File: src/LodeRunner/src/ValidationTest/Main.cs (hunk 5 above)
Comment (gortegaMS) [reply src/LodeRunner/src/ValidationTest/Main.cs:264]: Fixed

File: src/LodeRunner/src/ValidationTest/Main.cs (hunk 6 above)
Comment (gortegaMS) [reply src/LodeRunner/src/ValidationTest/Main.cs:403]: fixed

File: src/LodeRunner.Data/CosmosDBRepository.cs (hunk 3 above)
Comment (gortegaMS) [reply src/LodeRunner.Data/CosmosDBRepository.cs:207]: refactored

**Comment by gortegaMS** [comment]
After a conversation with the team we are taking a new approach , will close this PR and create a new one,

**Comment by gortegaMS** [comment]
This is no longer a valid PR , it was reopen to do some testing

**Pull request closed without merging**